File: 20-valid-parentheses/python/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        brackets = {"(" : ")", "{" : "}", "[" : "]"}
        stack = []
        for bracket in s:
            if bracket in brackets:
                stack.append(brackets[bracket])
            else:
                if not stack or stack.pop() != bracket:
                    return False
        return not stack

File: 20-valid-parentheses/golang/solution.go
package golang

func isValid(s string) bool {
	var stack []rune
	bracketPairs := map[rune]rune{}
	bracketPairs['('] = ')'
	bracketPairs['['] = ']'
	bracketPairs['{'] = '}'
	for _, bracket := range s {
		if closed, ok := bracketPairs[bracket]; ok {
			stack = append(stack, closed)
		} else {
			if len(stack) == 0 || stack[len(stack)-1] != bracket {
				return false
			}
			stack = stack[:len(stack)-1]
		}
	}
	return len(stack) == 0
}
